crates/compiler/src/obj_gen/util.rs

```diff
@@ -0,0 +1,265 @@
+        .get_operands()
+        .map(|operand| extract_value_operand(operand, opcode))
+        .collect::<Result<Vec<BasicValueEnum>>>()
+}
+
+/// Extracts pairs of `(name, type)` from the provided value operands.
+///
+/// # Errors
+///
+/// - [`Error::UnsupportedType`] if a scalable vector value is encountered in
+///   the provided `operands`.
+pub fn name_type_pairs_from_value_operands(
+    operands: &[BasicValueEnum],
+) -> Result<Vec<(String, LLVMType)>> {
+    operands
+        .iter()
+        .map(|bv| match bv {
+            BasicValueEnum::ArrayValue(array_val) => {
+                let ty = LLVMType::try_from(array_val.get_type())?;
+                let name = array_val.get_name().to_str()?.to_string();
+
+                Ok((name, ty))
+            }
+            BasicValueEnum::IntValue(int_val) => {
+                let ty = LLVMType::try_from(int_val.get_type())?;
+                let name = int_val.get_name().to_str()?.to_string();
+
+                Ok((name, ty))
+            }
+            BasicValueEnum::FloatValue(float_val) => {
+                let ty = LLVMType::try_from(float_val.get_type())?;
+                let name = float_val.get_name().to_str()?.to_string();
+
+                Ok((name, ty))
+            }
+            BasicValueEnum::PointerValue(ptr_val) => {
+                let ty = LLVMType::try_from(ptr_val.get_type())?;
+                let name = ptr_val.get_name().to_str()?.to_string();
+
+                Ok((name, ty))
+            }
+            BasicValueEnum::StructValue(struct_val) => {
+                let ty = LLVMType::try_from(struct_val.get_type())?;
+                let name = struct_val.get_name().to_str()?.to_string();
+
+                Ok((name, ty))
+            }
+            BasicValueEnum::VectorValue(vector_val) => {
+                Err(Error::UnsupportedType(vector_val.to_string()))?
+            }
+        })
+        .collect()
+}
+
+/// Extracts an integer from the provided basic `value`, or returns [`None`]
+/// if `value` is non-integral.
+#[must_use]
+pub fn int_from_bv(value: BasicValueEnum) -> Option<i64> {
+    match value {
+        BasicValueEnum::IntValue(int_val) => int_val.get_sign_extended_constant(),
+        _ => None,
+    }
+}
+
+/// Extracts an integer from the provided basic `value` where possible.
+///
+/// # Panics
+///
+/// If the provided `value` is not an integral value.
+#[must_use]
+pub fn expect_int_from_bv(value: BasicValueEnum) -> i64 {
+    int_from_bv(value).expect("The provided value was not an integer")
+}
+
+/// Extracts a name from the provided basic `value`, or returns [`None`] if
+/// that `value` is not possible.
+///
+/// # Errors
+///
+/// - [`Error::CStrConversionError`] if the name cannot be converted into a
+///   rust-native String.
+pub fn name_from_bv(value: BasicValueEnum) -> Result<Option<String>> {
+    let name = match value {
+        BasicValueEnum::PointerValue(ptr_val) => ptr_val.get_name().to_str()?.to_string(),
```

Comment:
nit: lots of duplication; prefer an inner function

Reply:
No way to make it work due to temporaries and differing types.